README.md
# Version Comparison

A Laravel package that compares the current git commit SHA with a remotely tracked version to see if the application is outdated, this is used in some of my projects to allow users to see when their instance of my software is out of date.

> **Note:** The project relies on private version tracking software, it will not do anything without a tracking ID the system can use to compare commit SHAs.

## Installation

You can install the package via composer:

```bash
composer require senither/version-comparison
```

You can publish the config file with:

```bash
php artisan vendor:publish --provider="Senither\VersionComparison\VersionComparisonServiceProvider" --tag="config"
```

This is the contents of the published config file:

```php
return [

    /*
    |--------------------------------------------------------------------------
    | Version Tracking ID
    |--------------------------------------------------------------------------
    |
    | The unique ID for the project that the latest git commit SHA
    | should be compared with via the API, this can be found on
    | the version tracker when selecting the project.
    |
    | @url https://vt.senither.com/dashboard
    */

    'id' => env('VERSION_COMPARISON_ID', null),

    /*
    |--------------------------------------------------------------------------
    | Cache Expiration Times
    |--------------------------------------------------------------------------
    |
    | The amount of time in seconds that should elapse before
    | the current commit hash is refreshed, or the latest
    | version is fetched from the API.
    |
    */

    'cache_time' => [
        'version' => 7200,
        'commit_hash' => 3600,
    ],
];

```

## Usage

```php
use Senither\VersionComparison\Facades\Version;

// Gets the current version
Version::getCurrentVersion();
```

## License

The Version Comparison package open-sourced software licensed under the [MIT license](LICENSE.md).


CHANGELOG.md
# Changelog

All notable changes to **Version Comparison** will be documented in this file.

## v1.1.0 - 2021-03-31

-   Move cache times to the config to make them customizable.

## v1.0.4 - 2021-03-31

-   Add support for PHP 8.0+

## v1.0.3 - 2021-03-30

-   Fix raw version being returned correctly.

## v1.0.2 - 2021-03-30

-   Fix facade being registered with the right namespace.
    -   I'm a big dumb dumb.

## v1.0.1 - 2021-03-30

-   Fix facade being registered with the right name.

## v1.0.0 - 2021-03-30

-   Initial release.


src/Version.php
<?php

namespace Senither\VersionComparison;

use Carbon\Carbon;

class Version
{
    /**
     * The JSON version response from the API.
     *
     * @var object
     */
    private $data;

    /**
     * Creates a new version instance using
     * the given JSON version response.
     *
     * @param object $data
     */
    public function __construct($data)
    {
        $this->data = $data;
    }

    /**
     * Check if there is a newere release.
     *
     * @return boolean
     */
    public function isOutdated(): bool
    {
        return $this->getVersionsBehind() > 0;
    }

    /**
     * Get the amount of versions the current version is behind.
     *
     * @return int
     */
    public function getVersionsBehind(): int
    {
        return $this->data->versions_behind;
    }

    /**
     * Get the commit hash.
     *
     * @return string
     */
    public function getHash(): string
    {
        return $this->data->hash;
    }

    /**
     * Get the commit message.
     *
     * @return string
     */
    public function getMessage(): string
    {
        return $this->data->message;
    }

    /**
     * Get the generated commit semantic version.
     *
     * @return string
     */
    public function getVersion(): string
    {
        return $this->data->version;
    }

    /**
     * Get the direct link to the commit URL.
     *
     * @return string
     */
    public function getCommitUrl(): string
    {
        return $this->data->url;
    }

    /**
     * Get the time the commit was made at.
     *
     * @return \Carbon\Carbon
     */
    public function getCommitTime(): Carbon
    {
        return new Carbon($this->data->commit_at);
    }

    /**
     * Return the semantic version when the class is cast to a string.
     *
     * @return string
     */
    public function __toString()
    {
        return $this->getVersion();
    }
}


src/VersionManager.php
<?php

namespace Senither\VersionComparison;

use ErrorException;

class VersionManager
{
    /**
     * Get the current version from the version tracker, or null if
     * no git SHA is found, or the API responds with an error.
     *
     * @return \Senither\VersionComparison\Version|null
     */
    public function getCurrentVersion(): ?Version
    {
        $projectId = config('version-comparison.id');
        if ($projectId == null) {
            return null;
        }

        $latestCommitHash = $this->getLatestCommitHash();
        if ($latestCommitHash == null) {
            return null;
        }

        return cache()->remember(
            'version-manger::commit-' . $latestCommitHash,
            config('version-comparison.cache_time.version'),
            fn () => $this->fetchCommitFromApi($projectId, $latestCommitHash)
        );
    }

    /**
     * Get the latest commit hash.
     *
     * @return string|null
     */
    protected function getLatestCommitHash(): ?string
    {
        $commitHash = cache()->remember(
            'version-manger::latest-commit-hash',
            config('version-comparison.cache_time.commit_hash'),
            fn () => exec('git rev-parse HEAD')
        );

        if (mb_strlen($commitHash) !== 40) {
            return null;
        }

        return $commitHash;
    }

    /**
     * Fetch the commit object from the API using
     * the given project ID and commit hash.
     *
     * @param  string $projectId
     * @param  string $commitHash
     * @return \Senither\VersionComparison\Version|null
     */
    protected function fetchCommitFromApi($projectId, $commitHash): ?Version
    {
        try {
            $responses = file_get_contents(
                'https://vt.senither.com/api/repo/' . $projectId . '/commits/' . $commitHash
            );

            return new Version(json_decode($responses));
        } catch (ErrorException $e) {
            return null;
        }
    }
}


src/VersionComparisonServiceProvider.php
<?php

namespace Senither\VersionComparison;

use Illuminate\Support\ServiceProvider;

class VersionComparisonServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap any package services.
     *
     * @return void
     */
    public function boot()
    {
        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__ . '/../config/version-comparison.php' => config_path('version-comparison.php'),
            ], 'config');
        }
    }

    /**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        $this->app->singleton('version', function () {
            return new VersionManager;
        });

        $this->mergeConfigFrom(__DIR__ . '/../config/version-comparison.php', 'version-comparison');
    }
}


config/version-comparison.php
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Version Tracking ID
    |--------------------------------------------------------------------------
    |
    | The unique ID for the project that the latest git commit SHA
    | should be compared with via the API, this can be found on
    | the version tracker when selecting the project.
    |
    | @url https://vt.senither.com/dashboard
    */

    'id' => env('VERSION_COMPARISON_ID', null),

    /*
    |--------------------------------------------------------------------------
    | Cache Expiration Times
    |--------------------------------------------------------------------------
    |
    | The amount of time in seconds that should elapse before
    | the current commit hash is refreshed, or the latest
    | version is fetched from the API.
    |
    */

    'cache_time' => [
        'version' => 7200,
        'commit_hash' => 3600,
    ],
];
